File: ChessLab/Maia/OpponentStyle.swift
import ChessKit
import Foundation

/// Un trait de coup, lu en rejouant le coup sur un plateau : ce qu'un
/// personnage AIME ou ÉVITE. Chaque trait vaut entre −1 et 1 (souvent 0 ou 1).
enum StyleTrait: String, CaseIterable, Codable, Hashable, Sendable {
    /// Le coup donne échec.
    case check
    /// Le coup capture.
    case capture
    /// Gain de matériel net (capture moins risque immédiat), ramené à ±1.
    case materialGain
    /// La pièce jouée peut être prise par une pièce de valeur nettement
    /// inférieure sans que le coup ne rapporte autant : un sacrifice, ou une
    /// pièce laissée en prise.
    case sacrifice
    /// Une pièce (ni pion ni roi) se rapproche du roi adverse.
    case towardKing
    /// Un pion avance sur l'aile où le roi adverse a roqué.
    case pawnStorm
    /// Roque.
    case castle
    /// Une pièce mineure quitte sa case de départ.
    case development
    /// Pions isolés ou doublés créés (positif) ou résorbés (négatif).
    case weakPawn
    /// Échange de pièces de même valeur (au moins une pièce mineure).
    case equalTrade
    /// Échange des dames.
    case queenTrade
    /// Captures nouvellement offertes à l'adversaire : la position se complique.
    case tension
    /// Mobilité gagnée par la pièce jouée.
    case mobility
}

/// Le style d'un personnage : un poids par trait, et une borne.
///
/// Le score d'un coup est la somme pondérée de ses traits, BORNÉE à
/// ±`strength` (en nats) : la probabilité que Maia lui attribue est
/// multipliée par e^score. Avec `strength` = 1, un coup adoré pèse au plus
/// 2,7 fois plus, un coup détesté 2,7 fois moins — le style colore la
/// distribution humaine, il ne la remplace pas, et il n'achète jamais une
/// gaffe que Maia jugeait improbable.
struct StyleProfile: Hashable, Sendable {
    var weights: [StyleTrait: Double]
    var strength: Double

    static let none = StyleProfile(weights: [:], strength: 0)

    var isNeutral: Bool { strength <= 0 || weights.isEmpty }

    /// Le même style, avec des poids ajoutés (style dynamique selon le score).
    func adding(_ extra: [StyleTrait: Double]) -> StyleProfile {
        guard !extra.isEmpty else { return self }
        var merged = weights
        for (trait, weight) in extra { merged[trait, default: 0] += weight }
        return StyleProfile(weights: merged, strength: max(strength, 0.6))
    }
}

/// Les traits d'UN coup, mesurés.
struct MoveTraits: Equatable, Sendable {
    var check = false
    var capture = false
    var capturedValue = 0
    var movedValue = 0
    var sacrifice = false
    var towardKing = false
    var pawnStorm = false
    var castle = false
    var development = false
    var weakPawnDelta = 0
    var equalTrade = false
    var queenTrade = false
    var tensionDelta = 0
    var mobilityDelta = 0

    /// Valeur du trait entre −1 et 1.
    func value(of trait: StyleTrait) -> Double {
        switch trait {
        case .check: check ? 1 : 0
        case .capture: capture ? 1 : 0
        case .materialGain: Self.clamp(Double(capturedValue - (sacrifice ? movedValue : 0)) / 5)
        case .sacrifice: sacrifice ? 1 : 0
        case .towardKing: towardKing ? 1 : 0
        case .pawnStorm: pawnStorm ? 1 : 0
        case .castle: castle ? 1 : 0
        case .development: development ? 1 : 0
        case .weakPawn: Self.clamp(Double(weakPawnDelta) / 2)
        case .equalTrade: equalTrade ? 1 : 0
        case .queenTrade: queenTrade ? 1 : 0
        case .tension: Self.clamp(Double(tensionDelta) / 3)
        case .mobility: Self.clamp(Double(mobilityDelta) / 6)
        }
    }

    private static func clamp(_ x: Double) -> Double { min(1, max(-1, x)) }
}

/// Extraction des traits et repondération de la distribution de Maia.
/// Module pur : ChessKit seulement, testé sur cas choisis.
enum OpponentStyle {

    static func value(of kind: Piece.Kind) -> Int {
        switch kind {
        case .pawn: 1
        case .knight, .bishop: 3
        case .rook: 5
        case .queen: 9
        case .king: 0
        }
    }

    /// Les traits du coup `lan` (UCI) pour le camp au trait, ou `nil` si le
    /// coup n'est pas applicable.
    static func traits(of lan: String, on board: Board) -> MoveTraits? {
        guard lan.count >= 4 else { return nil }
        let mover = board.position.sideToMove
        let from = Square(String(lan.prefix(2)))
        let to = Square(String(lan.dropFirst(2).prefix(2)))
        guard let piece = board.position.piece(at: from), piece.color == mover else { return nil }

        var scratch = board
        guard let made = scratch.move(pieceAt: from, to: to) else { return nil }
        var promotedKind: Piece.Kind?
        if case .promotion = scratch.state {
            let kind = lan.count == 5 ? (Piece.Kind(rawValue: String(lan.suffix(1)).uppercased()) ?? .queen) : .queen
            scratch.completePromotion(of: made, to: kind)
            promotedKind = kind
        }

        var traits = MoveTraits()
        traits.movedValue = value(of: promotedKind ?? piece.kind)
        if case let .capture(captured) = made.result {
            traits.capture = true
            traits.capturedValue = value(of: captured.kind)
            traits.queenTrade = captured.kind == .queen && piece.kind == .queen
        }
        switch scratch.state {
        case .check, .checkmate: traits.check = true
        default: break
        }
        traits.castle = piece.kind == .king && abs(from.file.number - to.file.number) == 2
        let homeRank = mover == .white ? 1 : 8
        traits.development = (piece.kind == .knight || piece.kind == .bishop)
            && from.rank.value == homeRank && to.rank.value != homeRank

        let enemy = mover.opposite
        if let king = board.position.pieces.first(where: { $0.kind == .king && $0.color == enemy }) {
            if piece.kind != .pawn, piece.kind != .king {
                let before = distance(from, king.square)
                let after = distance(to, king.square)
                traits.towardKing = after < before && after <= 3
            }
            if piece.kind == .pawn {
                let kingFile = king.square.file.number
                let kingWing: Int? = kingFile <= 3 ? 0 : (kingFile >= 6 ? 1 : nil)
                let pawnWing: Int? = to.file.number <= 3 ? 0 : (to.file.number >= 6 ? 1 : nil)
                let relativeRank = mover == .white ? to.rank.value : 9 - to.rank.value
                traits.pawnStorm = kingWing != nil && kingWing == pawnWing && relativeRank >= 4
            }
        }

        // Attaquants adverses de la case d'arrivée, APRÈS le coup.
        // `legalMoves(forPieceAt:)` ignore le trait : on peut interroger les
        // pièces adverses directement.
        var lowestAttacker: Int?
        for enemyPiece in scratch.position.pieces where enemyPiece.color == enemy {
            if scratch.legalMoves(forPieceAt: enemyPiece.square).contains(to) {
                let attackerValue = value(of: enemyPiece.kind)
                lowestAttacker = min(lowestAttacker ?? Int.max, attackerValue)
            }
        }
        if let lowestAttacker {
            let netCapture = traits.capture && traits.capturedValue >= traits.movedValue
            traits.sacrifice = lowestAttacker + 1 < traits.movedValue && !netCapture
            traits.equalTrade = traits.capture && traits.capturedValue == traits.movedValue && traits.movedValue >= 3
        }

        traits.weakPawnDelta = weakPawns(in: scratch.position, of: mover) - weakPawns(in: board.position, of: mover)
        traits.tensionDelta = captures(available: enemy, in: scratch) - captures(available: enemy, in: board)
        let mobilityBefore = board.legalMoves(forPieceAt: from).count
        let mobilityAfter = scratch.legalMoves(forPieceAt: to).count
        traits.mobilityDelta = mobilityAfter - mobilityBefore
        return traits
    }

    /// Score de style d'un coup, borné à ±`strength`.
    static func score(_ traits: MoveTraits, style: StyleProfile) -> Double {
        guard !style.isNeutral else { return 0 }
        let raw = style.weights.reduce(0.0) { $0 + $1.value * traits.value(of: $1.key) }
        return min(style.strength, max(-style.strength, raw))
    }

    /// Repondère les `topK` premiers candidats de Maia par leur score de
    /// style, puis renormalise. L'ordre de sortie est celui des probabilités
    /// repondérées.
    static func apply(_ style: StyleProfile, to candidates: [MaiaCandidate], board: Board, topK: Int = 8) -> [MaiaCandidate] {
        guard !style.isNeutral, !candidates.isEmpty else { return candidates }
        var weighted = candidates
        for index in 0..<min(topK, candidates.count) {
            guard let traits = traits(of: candidates[index].move.uci, on: board) else { continue }
            let factor = exp(score(traits, style: style))
            weighted[index] = MaiaCandidate(move: candidates[index].move, probability: candidates[index].probability * factor)
        }
        let total = weighted.reduce(0.0) { $0 + $1.probability }
        guard total > 0 else { return candidates }
        return weighted
            .map { MaiaCandidate(move: $0.move, probability: $0.probability / total) }
            .sorted { $0.probability > $1.probability }
    }

    // MARK: Outils

    static func distance(_ a: Square, _ b: Square) -> Int {
        max(abs(a.file.number - b.file.number), abs(a.rank.value - b.rank.value))
    }

    /// Pions isolés + pions doublés d'un camp.
    static func weakPawns(in position: Position, of color: Piece.Color) -> Int {
        var perFile = [Int](repeating: 0, count: 10)   // index 1...8, marges à 0 et 9
        for pawn in position.pieces where pawn.color == color && pawn.kind == .pawn {
            perFile[pawn.square.file.number] += 1
        }
        var weak = 0
        for file in 1...8 where perFile[file] > 0 {
            if perFile[file - 1] == 0, perFile[file + 1] == 0 { weak += perFile[file] }
            if perFile[file] > 1 { weak += perFile[file] - 1 }
        }
        return weak
    }

    /// Captures disponibles pour `color` (le trait est ignoré par ChessKit).
    static func captures(available color: Piece.Color, in board: Board) -> Int {
        var count = 0
        for piece in board.position.pieces where piece.color == color {
            for target in board.legalMoves(forPieceAt: piece.square) {
                if let victim = board.position.piece(at: target), victim.color != color { count += 1 }
            }
        }
        return count
    }
}

extension OpponentProfile {
    /// Le personnage tel qu'il joue MAINTENANT : température et style
    /// modulés par le score de la partie (point de vue du personnage).
    struct Mood: Equatable, Sendable {
        let temperature: Double
        let style: StyleProfile
    }

    func mood(lastMoverCp: Int?) -> Mood {
        guard let cp = lastMoverCp else { return Mood(temperature: temperature, style: style) }
        if cp >= Temperament.winningThresholdCp {
            return Mood(temperature: temperament.temperatureWhenWinning ?? temperature,
                        style: style.adding(temperament.styleWhenWinning))
        }
        if cp <= -Temperament.winningThresholdCp {
            return Mood(temperature: temperament.temperatureWhenLosing ?? temperature,
                        style: style.adding(temperament.styleWhenLosing))
        }
        return Mood(temperature: temperature, style: style)
    }
}

File: android/maia/src/main/java/com/chesslab/maia/OpponentStyle.kt
package com.chesslab.maia

import chesskit.Board
import chesskit.Move
import chesskit.Piece
import chesskit.Position
import chesskit.Square
import kotlin.math.abs
import kotlin.math.exp
import kotlin.math.max
import kotlin.math.min

/**
 * Traduction Kotlin de `OpponentStyle.swift`.
 *
 * Un trait de coup, lu en REJOUANT le coup sur un plateau : ce qu'un
 * personnage aime ou évite. Chaque trait vaut entre −1 et 1.
 */
@Suppress("EnumEntryName")
enum class StyleTrait {
    check, capture, materialGain, sacrifice, towardKing, pawnStorm,
    castle, development, weakPawn, equalTrade, queenTrade, tension, mobility,
}

/**
 * Le style d'un personnage : un poids par trait, et une borne.
 *
 * Le score d'un coup est la somme pondérée de ses traits, BORNÉE à
 * ±[strength] (en nats) : la probabilité de Maia est multipliée par e^score.
 * Avec `strength` = 1, un coup adoré pèse au plus 2,7 fois plus — le style
 * COLORE la distribution humaine, il ne la remplace pas, et il n'achète
 * jamais une gaffe que Maia jugeait improbable.
 */
data class StyleProfile(val weights: Map<StyleTrait, Double>, val strength: Double) {
    val isNeutral: Boolean get() = strength <= 0 || weights.isEmpty()

    /** Le même style, avec des poids ajoutés — le style dynamique selon le score. */
    fun adding(extra: Map<StyleTrait, Double>): StyleProfile {
        if (extra.isEmpty()) return this
        val merged = weights.toMutableMap()
        for ((trait, weight) in extra) merged[trait] = (merged[trait] ?: 0.0) + weight
        return StyleProfile(merged, max(strength, 0.6))
    }

    companion object { val none = StyleProfile(emptyMap(), 0.0) }
}

/** Les traits d'UN coup, mesurés. */
data class MoveTraits(
    var check: Boolean = false,
    var capture: Boolean = false,
    var capturedValue: Int = 0,
    var movedValue: Int = 0,
    var sacrifice: Boolean = false,
    var towardKing: Boolean = false,
    var pawnStorm: Boolean = false,
    var castle: Boolean = false,
    var development: Boolean = false,
    var weakPawnDelta: Int = 0,
    var equalTrade: Boolean = false,
    var queenTrade: Boolean = false,
    var tensionDelta: Int = 0,
    var mobilityDelta: Int = 0,
) {
    fun value(trait: StyleTrait): Double = when (trait) {
        StyleTrait.check -> if (check) 1.0 else 0.0
        StyleTrait.capture -> if (capture) 1.0 else 0.0
        StyleTrait.materialGain -> clamp((capturedValue - (if (sacrifice) movedValue else 0)) / 5.0)
        StyleTrait.sacrifice -> if (sacrifice) 1.0 else 0.0
        StyleTrait.towardKing -> if (towardKing) 1.0 else 0.0
        StyleTrait.pawnStorm -> if (pawnStorm) 1.0 else 0.0
        StyleTrait.castle -> if (castle) 1.0 else 0.0
        StyleTrait.development -> if (development) 1.0 else 0.0
        StyleTrait.weakPawn -> clamp(weakPawnDelta / 2.0)
        StyleTrait.equalTrade -> if (equalTrade) 1.0 else 0.0
        StyleTrait.queenTrade -> if (queenTrade) 1.0 else 0.0
        StyleTrait.tension -> clamp(tensionDelta / 3.0)
        StyleTrait.mobility -> clamp(mobilityDelta / 6.0)
    }

    private fun clamp(x: Double) = min(1.0, max(-1.0, x))
}

/** Extraction des traits et repondération de la distribution de Maia. */
object OpponentStyle {

    fun value(kind: Piece.Kind): Int = when (kind) {
        Piece.Kind.pawn -> 1
        Piece.Kind.knight, Piece.Kind.bishop -> 3
        Piece.Kind.rook -> 5
        Piece.Kind.queen -> 9
        Piece.Kind.king -> 0
    }

    /**
     * Les traits du coup [lan] pour le camp au trait, ou `null` si le coup
     * n'est pas applicable.
     *
     * L'original profite de la sémantique de valeur de Swift (`var scratch =
     * board`) ; ici le plateau d'essai est construit sur une COPIE explicite
     * de la position.
     */
    fun traits(lan: String, board: Board): MoveTraits? {
        if (lan.length < 4) return null
        val mover = board.position.sideToMove
        val from = Square(lan.substring(0, 2))
        val to = Square(lan.substring(2, 4))
        val piece = board.position.piece(from) ?: return null
        if (piece.color != mover) return null

        val scratch = Board(board.position.copy())
        var made: Move = scratch.move(pieceAt = from, to = to) ?: return null
        var promotedKind: Piece.Kind? = null
        if (scratch.state is Board.State.Promotion) {
            val kind = if (lan.length == 5) kindOf(lan[4]) else Piece.Kind.queen
            made = scratch.completePromotion(of = made, to = kind)
            promotedKind = kind
        }

        val traits = MoveTraits()
        traits.movedValue = value(promotedKind ?: piece.kind)
        (made.result as? Move.Result.Capture)?.let { captured ->
            traits.capture = true
            traits.capturedValue = value(captured.piece.kind)
            traits.queenTrade = captured.piece.kind == Piece.Kind.queen && piece.kind == Piece.Kind.queen
        }
        traits.check = scratch.state is Board.State.Check || scratch.state is Board.State.Checkmate
        traits.castle = piece.kind == Piece.Kind.king && abs(from.file.number - to.file.number) == 2

        val homeRank = if (mover == Piece.Color.white) 1 else 8
        traits.development = (piece.kind == Piece.Kind.knight || piece.kind == Piece.Kind.bishop) &&
            from.rank.value == homeRank && to.rank.value != homeRank

        val enemy = mover.opposite
        board.position.pieces.firstOrNull { it.kind == Piece.Kind.king && it.color == enemy }?.let { king ->
            if (piece.kind != Piece.Kind.pawn && piece.kind != Piece.Kind.king) {
                val before = distance(from, king.square)
                val after = distance(to, king.square)
                traits.towardKing = after < before && after <= 3
            }
            if (piece.kind == Piece.Kind.pawn) {
                val kingFile = king.square.file.number
                val kingWing = if (kingFile <= 3) 0 else if (kingFile >= 6) 1 else null
                val pawnWing = if (to.file.number <= 3) 0 else if (to.file.number >= 6) 1 else null
                val relativeRank = if (mover == Piece.Color.white) to.rank.value else 9 - to.rank.value
                traits.pawnStorm = kingWing != null && kingWing == pawnWing && relativeRank >= 4
            }
        }

        // les attaquants adverses de la case d'arrivée, APRÈS le coup :
        // `legalMoves` ignore le trait, on peut donc interroger l'adversaire
        var lowestAttacker: Int? = null
        for (enemyPiece in scratch.position.pieces) {
            if (enemyPiece.color != enemy) continue
            if (to in scratch.legalMoves(forPieceAt = enemyPiece.square)) {
                val attackerValue = value(enemyPiece.kind)
                lowestAttacker = min(lowestAttacker ?: Int.MAX_VALUE, attackerValue)
            }
        }
        lowestAttacker?.let { lowest ->
            val netCapture = traits.capture && traits.capturedValue >= traits.movedValue
            traits.sacrifice = lowest + 1 < traits.movedValue && !netCapture
            traits.equalTrade = traits.capture && traits.capturedValue == traits.movedValue && traits.movedValue >= 3
        }

        traits.weakPawnDelta = weakPawns(scratch.position, mover) - weakPawns(board.position, mover)
        traits.tensionDelta = captures(enemy, scratch) - captures(enemy, board)
        traits.mobilityDelta = scratch.legalMoves(forPieceAt = to).size - board.legalMoves(forPieceAt = from).size
        return traits
    }

    /** Le score de style d'un coup, borné à ±`strength`. */
    fun score(traits: MoveTraits, style: StyleProfile): Double {
        if (style.isNeutral) return 0.0
        val raw = style.weights.entries.sumOf { it.value * traits.value(it.key) }
        return min(style.strength, max(-style.strength, raw))
    }

    /**
     * Repondère les [topK] premiers candidats de Maia par leur score de style,
     * puis renormalise.
     *
     * Seulement les premiers : calculer les traits coûte un plateau d'essai
     * par coup, et la queue de la distribution ne pèse rien.
     */
    fun apply(style: StyleProfile, candidates: List<MaiaCandidate>, board: Board, topK: Int = 8): List<MaiaCandidate> {
        if (style.isNeutral || candidates.isEmpty()) return candidates
        val weighted = candidates.toMutableList()
        for (index in 0 until min(topK, candidates.size)) {
            val traits = traits(candidates[index].move.uci, board) ?: continue
            val factor = exp(score(traits, style))
            weighted[index] = candidates[index].copy(probability = candidates[index].probability * factor)
        }
        val total = weighted.sumOf { it.probability }
        if (total <= 0) return candidates
        return weighted.map { it.copy(probability = it.probability / total) }
            .sortedByDescending { it.probability }
    }

    fun distance(a: Square, b: Square): Int =
        max(abs(a.file.number - b.file.number), abs(a.rank.value - b.rank.value))

    /** Pions isolés + pions doublés d'un camp. */
    fun weakPawns(position: Position, color: Piece.Color): Int {
        val perFile = IntArray(10)      // 1..8 utilisés, marges à 0 et 9
        for (pawn in position.pieces) {
            if (pawn.color == color && pawn.kind == Piece.Kind.pawn) perFile[pawn.square.file.number]++
        }
        var weak = 0
        for (file in 1..8) {
            if (perFile[file] == 0) continue
            if (perFile[file - 1] == 0 && perFile[file + 1] == 0) weak += perFile[file]
            if (perFile[file] > 1) weak += perFile[file] - 1
        }
        return weak
    }

    /** Les captures disponibles pour `color` — le trait est ignoré. */
    fun captures(color: Piece.Color, board: Board): Int {
        var count = 0
        for (piece in board.position.pieces) {
            if (piece.color != color) continue
            for (target in board.legalMoves(forPieceAt = piece.square)) {
                val victim = board.position.piece(target)
                if (victim != null && victim.color != color) count++
            }
        }
        return count
    }

    private fun kindOf(c: Char): Piece.Kind = when (c.lowercaseChar()) {
        'r' -> Piece.Kind.rook
        'b' -> Piece.Kind.bishop
        'n' -> Piece.Kind.knight
        else -> Piece.Kind.queen
    }
}
